app/uk/gov/hmrc/helptosave/actors/EligibilityStatsHandler.scala

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2018 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.helptosave.actors
+
+import cats.instances.int._
+import cats.instances.option._
+import cats.instances.string._
+import cats.syntax.eq._
+import com.google.inject.{ImplementedBy, Singleton}
+import uk.gov.hmrc.helptosave.repo.MongoEligibilityStatsStore.EligibilityStats
+import uk.gov.hmrc.helptosave.util.Logging
+
+@ImplementedBy(classOf[EligibilityStatsHandlerImpl])
+trait EligibilityStatsHandler {
+  def handleStats(result: List[EligibilityStats]): String
+}
+
+@Singleton
+class EligibilityStatsHandlerImpl extends EligibilityStatsHandler with Logging {
+
+  override def handleStats(result: List[EligibilityStats]): String = {
```

Comment:
this is the original logic i wrote, but i have come up with the following for dynamically formatting but it looks an overkill, so lets keep the  logic as it as ? 

```
@Singleton
class EligibilityStatsHandlerImpl extends EligibilityStatsHandler with Logging {

  val knownSources = Set("Digital", "Stride", "KCOM", "Unknown")

  override def handleStats(result: List[EligibilityStats]): String = {

    val groupedByEligibilityReason: Map[Option[Int], List[EligibilityStats]] = result.groupBy(_.eligibilityReason)
    val groupedByEligibilityReasonThenSource: List[(String, ArrayBuffer[(String, Int)])] =
      groupedByEligibilityReason.map {
        case (k, v) ⇒
          k.map(_.toString).getOrElse("Unknown") -> {
            val z = v.groupBy(_.source).map(x ⇒ x._1.getOrElse("Unknown") → x._2.map(_.total).sum).toList.sortBy(_._1)
            ArrayBuffer.concat(z)
          }
      }.toList.sortBy(_._1)

    //fill totals of missing sources with 0
    knownSources.foreach {
      source ⇒
        groupedByEligibilityReasonThenSource.foreach {
          case (p1, p2) ⇒
            var ml = mutable.ArrayBuffer.empty[(String, Int)]
            if (!p2.map(_._1).contains(source)) {
              val qq: (String, Int) = (source, 0)
              p2 += qq
            }
        }
    }

    //map integers to 6 ->  UC, 7 ->  WTC, 8 -> UC&WTC

    val newGroup = ArrayBuffer.empty[(String, ArrayBuffer[(String, Int)])]

    ArrayBuffer.concat(groupedByEligibilityReasonThenSource)
      .foreach {
        p ⇒
          if (p._1 === "6") {
            val t: (String, ArrayBuffer[(String, Int)]) = ("UC", ArrayBuffer.concat(p._2))
            newGroup += t
          } else if (p._1 === "7") {
            val t: (String, ArrayBuffer[(String, Int)]) = ("WTC", ArrayBuffer.concat(p._2))
            newGroup += t
          } else if (p._1 === "8") {
            val t: (String, ArrayBuffer[(String, Int)]) = ("UC&WTC", ArrayBuffer.concat(p._2))
            newGroup += t
          } else if (p._1 === "Unknown") {
            val t: (String, ArrayBuffer[(String, Int)]) = ("Unknown", ArrayBuffer.concat(p._2))
            newGroup += t
          }

      }

    logger.info(s"newgroup = $newGroup")

    val k = "|%8s|%10s|%10s|\n"

    val header: String =
      s"""
         |+--------+----------+----------+
         || Reason | Channel  |  Count   |
         |+--------+----------+----------+\n""".stripMargin

    var report = newGroup.foldLeft[String](header) {
      case (row: String, element: (String, ArrayBuffer[(String, Int)])) ⇒
        val a = element._2.foldLeft[String](row) {
          case (x: String, y: (String, Int)) ⇒
            x.concat(k.format(element._1, y._1, y._2))
        }

        def getTotalByReason =
          newGroup
            .filter(p ⇒ p._1 === element._1)
            .flatMap(_._2)
            .map(_._2)
            .sum

        a.concat(k.format("", "Total", getTotalByReason))
          .concat(k.format("", "", ""))

    }

    //compute totals by source
    knownSources.foreach {
      p =>
        val t = newGroup.map(_._2).map(_.toList.filter(_._1 === p).map(_._2)).map(_.sum).sum
        report = report.concat(k.format("Total", p, t))
    }

    //add grand total
    report = report
      .concat(k.format("", "Total", newGroup.map(_._2.map(_._2).sum).sum))
      .concat("+--------+----------+----------+")

    logger.info(s"report is $report")

    //for unit testing
    report
  }
```

Reply:
I'd still recommend using a generic approach - when we add in new channels we don't want to keep remembering to update this report, and if there is a bug and there is some data written with a reason or channel we don't expect we would want to see it. Looking at the generic implementation above, it could perhaps be simplified by breaking our some of the logic into different methods. I notice there is also a lot of mutable state in there - we can avoid mutable state here by using folds and/or (tail) recursive functions to do the looping logic.

Perhaps to make it simpler, it would be easier to factor out the logic into something which looks like this:
```scala
// inner list is to support sub-keys and their values
type Table = List[(String,List[(String,String)])]

// handles gathering stats
def createTable(state: List[EligibilityStats]): Table

// handlers generic formatting logic
def prettyFormatTable(table: Table): String = ??? 
```
Then the logic would look something like:
```
val table: Table = // build table
val report: String = prettyFormatTable(table)
```
That way it's more easily understood what's going on

Reply:
hm ok, just pushed the change, please check if anything can be improved